### PR TITLE
[ONNX] Add ONNX ExportedProgram tests

### LINKED ISSUE #113705: [ONNX] Track dynamic shapes integration for torch.onnx.dynamo_export
### 🐛 Describe the bug

- [x] Add meaningful dynamic shapes tests to guard `ExportedProgram` exported in `torch.onnx.dynamo_export`
- [x] https://github.com/pytorch/pytorch/issues/114166

### Versions

main branch

```[tasklist]
### Tasks
- [ ] https://github.com/pytorch/pytorch/issues/114954
```

### LINKED ISSUE #114166: [ONNX] Move torch.export.export tests out from `test_fx_to_onnx_onnxruntime.py`
Move the tests like https://github.com/pytorch/pytorch/blob/0c450f4504d1ca1fed0958f5454e393664ce9637/test/onnx/test_fx_to_onnx_with_onnxruntime.py#L899-L909

out from `test_fx_to_onnx_onnxruntime` to further separate `torch.onnx.dynamo_export` usages on nn.Module and ExportedProgram.

Also, it's a bug that the test with prior `torch.export.export` call while we already have another one inside `self.run_test_with_fx_to_onnx_exporter_and_onnx_runtime`. 

The idea of ExportedProgram extension tests can be kept in the test file.

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114633

Fix #114166 
Fix #113705 

This PR references tests from `test_export.py` to make sure the exported program from PyTorch can all be successfully exported into ONNX model.

### CODE REVIEW

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py [hunk 1]
```diff
@@ -43,8 +43,14 @@ def _parameterized_class_attrs_and_values():
...
-            (True, False),
-            (True, False),
+            (
+                True,
+                False,
+            ),
+            (
+                True,
+                False,
+            ),
```
Comment (titaiwangms) [inline test/onnx/test_fx_to_onnx_with_onnxruntime.py:53]: Set this back

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py [hunk 2]
```diff
@@ -181,35 +187,33 @@ def func(x, y):
...
-        ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(func(tensor_x, 9.0))
-        _ = onnx_test_common.run_ort(onnx_program, onnx_format_args)
-        for ref_output, ort_output in zip(ref_outputs, ort_outputs):
-            torch.testing.assert_close(ref_output, torch.tensor(ort_output))
+        # # test on different non-tensor input - xfail
+        # onnx_format_args = onnx_program.adapt_torch_inputs_to_onnx(tensor_x, 9.0)
+        # ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(func(tensor_x, 9.0))
+        # _ = onnx_test_common.run_ort(onnx_program, onnx_format_args)
+        # for ref_output, ort_output in zip(ref_outputs, ort_outputs):
+        #     torch.testing.assert_close(ref_output, torch.tensor(ort_output))
```
Comment (titaiwangms) [inline test/onnx/test_fx_to_onnx_with_onnxruntime.py:216]: Revert this back

File: torch/onnx/_internal/exporter.py [hunk 3]
```diff
@@ -1129,6 +1129,8 @@ def export(self) -> ONNXProgram:
                 self.options, self.model, self.model_args, self.model_kwargs
             )
 
+            graph_module.print_readable()
```
Comment (titaiwangms) [inline torch/onnx/_internal/exporter.py:1132]: Remove this

File: torch/onnx/_internal/io_adapter.py [hunk 4]
```diff
@@ -414,6 +414,7 @@ def apply(
                 is not identical to the `SpecTree` output produced from the first
                 `model_outputs` that was passed to this method.
         """
+        # import pdb; pdb.set_trace()
```
Comment (titaiwangms) [inline torch/onnx/_internal/io_adapter.py:417]: Remove this

File: test/onnx/torch_export/test_torch_export_with_onnxruntime.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+            count += 1
+            return None
+
+        dynamic_shapes = torch_pytree.tree_map(dynamify_inp, inp)
+        exported_program = torch.export.export(foo, inp, dynamic_shapes=dynamic_shapes)
+        onnx_program = torch.onnx.dynamo_export(exported_program, inp_a, inp_b)
+
+        # NOTE: Careful with the input format. The input format should be
+        # consistent with how the model is exported.
+        test_inputs = ((torch.randn(4, 4), torch.randn(6, 4)), torch.randn(4, 4))
```
Comment (thiagocrepaldi) [inline test/onnx/torch_export/test_torch_export_with_onnxruntime.py:263]: you should use `onnx_program.adapt_torch_inputs_tyo_onnx()` to adapt the "normal torch input" to the format expected by the onnx model

File: test/onnx/torch_export/test_torch_export_with_onnxruntime.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+            {"x": torch.randn(3, 3), "y": torch.randn(3, 3)},
+            dynamic_shapes=dynamic_shapes,
+        )
+        onnx_program = torch.onnx.dynamo_export(
+            exported_program, x=torch.randn(3, 3), y=torch.randn(3, 3)
+        )
+
+        # NOTE: A model should be fed with the input formats that
+        # how the model is exported
+        inputs = {"x": torch.randn(6, 3), "y": torch.randn(6, 3)}
```
Comment (thiagocrepaldi) [inline test/onnx/torch_export/test_torch_export_with_onnxruntime.py:291]: same as above. use `onnx_program.adapt_torch_inputs_tyo_onnx()`

File: test/onnx/torch_export/test_torch_export_with_onnxruntime.py (hunk 5 above)
Comment (titaiwangms) [reply test/onnx/torch_export/test_torch_export_with_onnxruntime.py:263]: Not sure about this. The input is adpated inside onnx_program __call__.

File: test/onnx/torch_export/test_torch_export_with_onnxruntime.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+        input_args,
+        input_kwargs=None,
+        rtol=1e-03,
+        atol=1e-07,
+    ):
+        # avoid mutable default argument
+        if input_kwargs is None:
+            input_kwargs = {}
+
+        onnx_outputs = onnx_exported_program(*input_args, **input_kwargs)
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply test/onnx/torch_export/test_torch_export_with_onnxruntime.py:34]: Added

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py [hunk 8]
```diff
@@ -515,9 +515,14 @@ def forward(self, x):
             additional_test_inputs=[((x2,),)],
         )
 
-    @pytorch_test_common.xfail(
-        "[ONNXRuntimeError] : 1 : FAIL : Type Error: Type (tensor(float)) of output arg (copy) "
-        "of node (n0__4) does not match expected type (tensor(int64))"
+    @pytorch_test_common.xfail_if_model_type_is_exportedprogram(
+        "RuntimeError:"
+        " Found following user inputs located at [0] are mutated. This is currently banned in the aot_export workflow."
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply test/onnx/test_fx_to_onnx_with_onnxruntime.py]: Error message changed

File: test/onnx/torch_export/test_torch_export_with_onnxruntime.py (hunk 6 above)
Comment (titaiwangms) [reply test/onnx/torch_export/test_torch_export_with_onnxruntime.py:294]: The note is kept for torch exported program usage.

**Review by thiagocrepaldi — APPROVED** [review]

**Pull request closed without merging**